### PR TITLE
feat: add transaction filter

### PR DESCRIPTION
**Pull request by quake**
This PR is an implementation of  [transaction filter protocol](https://github.com/nervosnetwork/rfcs/pull/50), but need further improvement:
- Merkle tree proof generation is not ready, waiting for MMR PR and reuse the generation function.
- Sync/Relay protocol share the `Peers` information data, should refactor and move it to network protocol.
- Unify H256 ser/dser.

### CODE REVIEW

File: sync/src/synchronizer/block_fetcher.rs [hunk 1]
```diff
@@ -1,5 +1,5 @@
-use super::header_view::HeaderView;
 use crate::synchronizer::{BlockStatus, Synchronizer};
+use crate::types::HeaderView;
```
Comment (doitian) [inline sync/src/synchronizer/block_fetcher.rs]: I don't see that the mod `header_view` is removed, so now we have two different `HeaderVIew`?

Comment (quake) [reply sync/src/synchronizer/block_fetcher.rs]: ops, I removed mod ref but forgot to delete this file. will push it later.

File: sync/src/relayer/transaction_process.rs [hunk 2]
```diff
@@ -38,7 +38,15 @@ where
...
 
             for peer_id in self.nc.connected_peers() {
-                if peer_id != self.peer {
+                if peer_id != self.peer
+                    && self
+                        .relayer
+                        .peers()
+                        .transaction_filters
+                        .read()
+                        .get(&self.peer)
```
Comment (jjyr) [inline sync/src/relayer/transaction_process.rs]: ???? should be peer_id?

File: sync/src/synchronizer/filter_process.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    peer: PeerIndex,
+}
+
+impl<'a, CI> AddFilterProcess<'a, CI>
+where
+    CI: ChainIndex + 'a,
+{
+    pub fn new(
+        message: &'a AddFilter,
+        synchronizer: &'a Synchronizer<CI>,
```
Comment (jjyr) [inline sync/src/synchronizer/filter_process.rs]: should receive transaction_fitlers instead the whole Synchronizer?

File: src/cli/run_impl.rs [hunk 4]
```diff
@@ -61,6 +61,7 @@ pub fn run(setup: Setup) {
         chain_controller.clone(),
         shared.clone(),
         tx_pool_controller.clone(),
+        synchronizer.peers(),
```
Comment (jjyr) [inline src/cli/run_impl.rs]: I have notice Peers have been moved out from synchronizer module, here is better to move out the initialize of Peer from Synchronizer struct to keep logical consistent.

File: protocol/src/builder.rs [hunk 5]
```diff
@@ -340,6 +340,47 @@ impl<'a> SyncMessage<'a> {
...
+        fbb: &mut FlatBufferBuilder<'b>,
+        block: &Block,
+        transactions_index: &[usize],
+    ) -> WIPOffset<FilteredBlock<'b>> {
+        let transactions = transactions_index
+            .iter()
+            .map(|ti| {
+                let fbs_transaction = FbsTransaction::build(fbb, &block.commit_transactions()[*ti]);
+                let mut builder = IndexTransactionBuilder::new(fbb);
+                builder.add_index(*ti as u32);
```
[the remark this answers is not in the document]
Comment (jjyr) [reply protocol/src/builder.rs]: There're other protocol struct builders also use `usize` as the parameter, should be fixed in a standalone PR.

**Pull request merged**